File: UnityHVOF/Assets/Scripts/Infrastructure/Laser/TopHatSource.cs
// ============================================================================
// Top-Hat Laser Source — Implements ILaserSource.
//
// Pattern: Strategy — interchangeable with GaussianSource.
// Computes uniform intensity within the spot radius, zero outside.
//
// Port of: infrastructure/laser/tophat_source.py
// ============================================================================

using System;
using HVOFSim.Domain.Ports;

namespace HVOFSim.Infrastructure.Laser
{
    /// <summary>Top-hat (flat-top) beam profile intensity distribution.</summary>
    public sealed class TopHatSource : ILaserSource
    {
        public double[,] ComputeIntensity(
            double[,] xGrid, double[,] yGrid,
            double centerX, double centerY,
            double spotRadius, double power)
        {
            int nx = xGrid.GetLength(0);
            int ny = xGrid.GetLength(1);
            var intensity = new double[nx, ny];

            double uniformIntensity = power / (Math.PI * spotRadius * spotRadius);
            double r2Max = spotRadius * spotRadius;

            for (int i = 0; i < nx; i++)
            {
                for (int j = 0; j < ny; j++)
                {
                    double dx = xGrid[i, j] - centerX;
                    double dy = yGrid[i, j] - centerY;
                    double r2 = dx * dx + dy * dy;
                    intensity[i, j] = r2 <= r2Max ? uniformIntensity : 0.0;
                }
            }

            return intensity;
        }
    }
}

File: infrastructure/laser/tophat_source.py
"""
Top-Hat Laser Source — Flat-top beam profile.

Implements ILaserSource using uniform intensity inside the spot:
    I = P / (πw²) inside the spot, 0 outside.

Single Responsibility: Only computes top-hat intensity distribution.
"""

import numpy as np

from domain.ports import ILaserSource


class TopHatLaserSource(ILaserSource):
    """Strategy: Flat-top (top-hat) beam intensity profile."""

    def compute_intensity(
        self,
        x_grid: np.ndarray,
        y_grid: np.ndarray,
        center_x: float,
        center_y: float,
        spot_radius: float,
        power: float,
    ) -> np.ndarray:
        """
        Compute flat-top intensity: uniform inside spot, zero outside.

        I = P / (πw²) for r ≤ w, else 0
        """
        r_squared = (x_grid - center_x) ** 2 + (y_grid - center_y) ** 2
        uniform_intensity = power / (np.pi * spot_radius ** 2)
        mask = r_squared <= spot_radius ** 2
        return np.where(mask, uniform_intensity, 0.0)
